js/contentSearchUI.js

```diff
@@ -0,0 +1,898 @@
+   * @param {String} idPrefix
+   *        The IDs of elements created by the object will be prefixed with this
+   *        string.
+   */
+  function ContentSearchUIController(inputElement, tableParent, healthReportKey,
+    searchPurpose, idPrefix = "") {
+    this.input = inputElement;
+    this._idPrefix = idPrefix;
+    this._healthReportKey = healthReportKey;
+    this._searchPurpose = searchPurpose;
+
+    let tableID = idPrefix + "searchSuggestionTable";
+    this.input.autocomplete = "off";
+    this.input.setAttribute("aria-autocomplete", "true");
+    this.input.setAttribute("aria-controls", tableID);
+    tableParent.appendChild(this._makeTable(tableID));
+
+    this.input.addEventListener("keypress", this);
+    this.input.addEventListener("input", this);
+    this.input.addEventListener("focus", this);
+    this.input.addEventListener("blur", this);
+
+    gNewTab.registerListener("NewTab:ContentSearchService",
+      this._onContentSearchService.bind(this));
+
+    this._stickyInputValue = "";
+    this._hideSuggestions();
+
+    this._getSearchEngines();
+    this._getStrings();
+  }
+
+  ContentSearchUIController.prototype = {
+
+    // The timeout (ms) of the remote suggestions.  Corresponds to
+    // SearchSuggestionController.remoteTimeout.  Uses
+    // SearchSuggestionController's default timeout if falsey.
+    remoteTimeout: undefined,
+    _oneOffButtons: [],
+    // Setting up the one off buttons causes an uninterruptible reflow. If we
+    // receive the list of engines while the newtab page is loading, this reflow
+    // may regress performance - so we set this flag and only set up the buttons
+    // if it's set when the suggestions table is actually opened.
+    _pendingOneOffRefresh: undefined,
+
+    get defaultEngine() {
+      return this._defaultEngine;
+    },
+
+    set defaultEngine(engine) {
+      let icon;
+      if (engine.iconBuffer) {
+        icon = this._getFaviconURIFromBuffer(engine.iconBuffer);
+      } else {
+        icon = this._getImageURIForCurrentResolution(
+          "chrome://mozapps/skin/places/defaultFavicon.png");
+      }
+      this._defaultEngine = {
+        name: engine.name,
+        icon: icon,
+      };
+      this._updateDefaultEngineHeader();
+
+      if (engine && document.activeElement === this.input) {
+        this._speculativeConnect();
+      }
+    },
+
+    get engines() {
+      return this._engines;
+    },
+
+    set engines(val) {
+      this._engines = val;
+      this._pendingOneOffRefresh = true;
+    },
+
+    // The selectedIndex is the index of the element with the "selected" class in
+    // the list obtained by concatenating the suggestion rows, one-off buttons, and
+    // search settings button.
+    get selectedIndex() {
+      let allElts = [...this._suggestionsList.children,
+        ...this._oneOffButtons,
+        document.getElementById("contentSearchSettingsButton")
+      ];
```

Comment:
heh, fancy... Maybe:
allElts = this._suggestionsList.children
  .concat(this._oneOffButtons)
  .concat(document.getElementById("contentSearchSettingsButton"));